File: go/server/network_check_test.go
package server

import (
	"errors"
	"strings"
	"testing"

	"github.com/solana-foundation/mpp-sdk/go"
)

// Pure-function tests for CheckNetworkBlockhash. The check is asymmetric:
// a Surfpool-prefixed blockhash is only valid on `localnet`, but a
// non-prefixed blockhash is accepted on any network (we can't tell from a
// non-prefixed hash which real cluster it came from).

// ── happy paths ───────────────────────────────────────────────────────────

func TestNetworkCheck_LocalnetWithSurfpoolHash_OK(t *testing.T) {
	if err := CheckNetworkBlockhash("localnet", "SURFNETxSAFEHASHxxxxxxxxxxxxxxxxxxx1892bcad"); err != nil {
		t.Fatalf("expected nil, got %v", err)
	}
}

func TestNetworkCheck_LocalnetWithRealHash_OK(t *testing.T) {
	// Real localnet validator (not Surfpool) — also valid.
	if err := CheckNetworkBlockhash("localnet", "11111111111111111111111111111111"); err != nil {
		t.Fatalf("expected nil, got %v", err)
	}
}

func TestNetworkCheck_MainnetWithRealHash_OK(t *testing.T) {
	if err := CheckNetworkBlockhash("mainnet", "9zrUHnA1nCByPksy3aL8tQ47vqdaG2vnFs4HrxgcZj4F"); err != nil {
		t.Fatalf("expected nil, got %v", err)
	}
}

func TestNetworkCheck_DevnetWithRealHash_OK(t *testing.T) {
	if err := CheckNetworkBlockhash("devnet", "EkSnNWid2cvwEVnVx9aBqawnmiCNiDgp3gUdkDPTKN1N"); err != nil {
		t.Fatalf("expected nil, got %v", err)
	}
}

// ── the actual bug surface ────────────────────────────────────────────────

func TestNetworkCheck_MainnetRejectsSurfpoolHash(t *testing.T) {
	err := CheckNetworkBlockhash("mainnet", "SURFNETxSAFEHASHxxxxxxxxxxxxxxxxxxx1892bcad")
	if err == nil {
		t.Fatal("expected error, got nil")
	}
	var sdkErr *mpp.Error
	if !errors.As(err, &sdkErr) {
		t.Fatalf("expected *mpp.Error, got %T", err)
	}
	if sdkErr.Code != mpp.ErrCodeWrongNetwork {
		t.Errorf("expected code %q, got %q", mpp.ErrCodeWrongNetwork, sdkErr.Code)
	}
	if !strings.Contains(sdkErr.Message, "Signed against localnet") {
		t.Errorf("missing received-side: %s", sdkErr.Message)
	}
	if !strings.Contains(sdkErr.Message, "server expects mainnet") {
		t.Errorf("missing expected-side: %s", sdkErr.Message)
	}
	if !strings.Contains(sdkErr.Message, "re-sign") {
		t.Errorf("missing actionable hint: %s", sdkErr.Message)
	}
}

func TestNetworkCheck_DevnetRejectsSurfpoolHash(t *testing.T) {
	err := CheckNetworkBlockhash("devnet", "SURFNETxSAFEHASHxxxxxxxxxxxxxxxxxxx1892bcad")
	if err == nil {
		t.Fatal("expected error, got nil")
	}
	if !strings.Contains(err.Error(), "server expects devnet") {
		t.Errorf("missing expected devnet: %s", err.Error())
	}
}

// ── edge cases ────────────────────────────────────────────────────────────

func TestNetworkCheck_PartialPrefixDoesNotMatch(t *testing.T) {
	// "SURFNETx" alone (8 chars) is NOT the full prefix.
	if err := CheckNetworkBlockhash("mainnet", "SURFNETx9zrUHnA1nCByPksy"); err != nil {
		t.Fatalf("expected nil, got %v", err)
	}
}

func TestNetworkCheck_ExactPrefixOnlyIsTreatedAsSurfpool(t *testing.T) {
	if err := CheckNetworkBlockhash("localnet", SurfpoolBlockhashPrefix); err != nil {
		t.Errorf("localnet+exact prefix should be ok: %v", err)
	}
	if err := CheckNetworkBlockhash("mainnet", SurfpoolBlockhashPrefix); err == nil {
		t.Error("mainnet+exact prefix should be wrong-network")
	}
}

func TestNetworkCheck_NonSurfpoolHashPassesAnywhere(t *testing.T) {
	// The check is asymmetric: a real-cluster-looking blockhash is
	// accepted on every network. Pin the design intent.
	for _, n := range []string{"mainnet", "devnet", "localnet"} {
		if err := CheckNetworkBlockhash(n, "11111111111111111111111111111111"); err != nil {
			t.Errorf("%s + real hash should be ok: %v", n, err)
		}
	}
}

File: typescript/packages/mpp/src/__tests__/network-check.test.ts
import { describe, expect, test } from 'vitest';

import {
    checkNetworkBlockhash,
    SURFPOOL_BLOCKHASH_PREFIX,
    WrongNetworkError,
} from '../server/network-check.js';

// Pure-function tests for the Surfpool-prefix-vs-non-localnet check.
// The check is asymmetric: a Surfpool-prefixed blockhash is only valid
// on `localnet`, but a non-prefixed blockhash is accepted on any network
// (we can't tell from a non-prefixed hash which real cluster it came from).

describe('checkNetworkBlockhash', () => {
    // ── happy paths ────────────────────────────────────────────────────────

    test('localnet + Surfpool-prefixed hash is ok', () => {
        expect(() =>
            checkNetworkBlockhash('localnet', 'SURFNETxSAFEHASHxxxxxxxxxxxxxxxxxxx1892bcad'),
        ).not.toThrow();
    });

    test('localnet + real hash is ok', () => {
        // Real localnet validator (not Surfpool) — also valid.
        expect(() =>
            checkNetworkBlockhash('localnet', '11111111111111111111111111111111'),
        ).not.toThrow();
    });

    test('mainnet + real hash is ok', () => {
        expect(() =>
            checkNetworkBlockhash('mainnet', '9zrUHnA1nCByPksy3aL8tQ47vqdaG2vnFs4HrxgcZj4F'),
        ).not.toThrow();
    });

    test('devnet + real hash is ok', () => {
        expect(() =>
            checkNetworkBlockhash('devnet', 'EkSnNWid2cvwEVnVx9aBqawnmiCNiDgp3gUdkDPTKN1N'),
        ).not.toThrow();
    });

    // ── the actual bug surface ─────────────────────────────────────────────

    test('mainnet rejects Surfpool-prefixed hash with explicit context', () => {
        let caught: unknown;
        try {
            checkNetworkBlockhash('mainnet', 'SURFNETxSAFEHASHxxxxxxxxxxxxxxxxxxx1892bcad');
        } catch (e) {
            caught = e;
        }
        expect(caught).toBeInstanceOf(WrongNetworkError);
        const err = caught as WrongNetworkError;
        expect(err.code).toBe('wrong-network');
        expect(err.expected).toBe('mainnet');
        expect(err.received).toBe('localnet');
        expect(err.message).toContain('Signed against localnet');
        expect(err.message).toContain('server expects mainnet');
        expect(err.message).toContain('re-sign');
        // Structured blockhash field is preserved for tooling even
        // though the message text doesn't include it.
        expect(err.blockhash).toBe('SURFNETxSAFEHASHxxxxxxxxxxxxxxxxxxx1892bcad');
    });

    test('devnet rejects Surfpool-prefixed hash', () => {
        expect(() =>
            checkNetworkBlockhash('devnet', 'SURFNETxSAFEHASHxxxxxxxxxxxxxxxxxxx1892bcad'),
        ).toThrow(WrongNetworkError);
    });

    // ── edge cases ─────────────────────────────────────────────────────────

    test('partial prefix (SURFNETx alone) does NOT match', () => {
        expect(() =>
            checkNetworkBlockhash('mainnet', 'SURFNETx9zrUHnA1nCByPksy'),
        ).not.toThrow();
    });

    test('exact prefix only is treated as Surfpool', () => {
        expect(() => checkNetworkBlockhash('localnet', SURFPOOL_BLOCKHASH_PREFIX)).not.toThrow();
        expect(() => checkNetworkBlockhash('mainnet', SURFPOOL_BLOCKHASH_PREFIX)).toThrow(
            WrongNetworkError,
        );
    });

    test('non-Surfpool hash passes on every network (asymmetric design)', () => {
        for (const network of ['mainnet', 'devnet', 'localnet']) {
            expect(() =>
                checkNetworkBlockhash(network, '11111111111111111111111111111111'),
            ).not.toThrow();
        }
    });
});
